### ISSUE TITLE
Docker Engine becomes unresponsive to some commands

### ISSUE THREAD

**Issue by soulshake** [moby/moby#35408]
## Summary

The Docker Engine becomes unresponsive to some commands (e.g. `docker info`, `docker pull`, `docker kill`) but others still work (e.g. `docker ps`, `docker version`). See **Command list** below for a more detailed list.

We haven't found a way to reproduce, but at any given moment, we find that about 20% of our EC2 instances (which tend to live between 24 and 48 hours) are affected by this issue.

So far, the only workaround we've found is to `restart docker`.

We are running 17.09.0-ce. For further configuration and version details, see **Context** at the end of this issue.

#### What I do

Run e.g. `docker info` or `curl 127.0.0.1:4243/v1.24/info`

#### What I expect to happen

To see `docker info` output.

#### What happens instead

The command hangs indefinitely:

```
# curl -vv 127.0.0.1:4243/v1.24/info
* Hostname was NOT found in DNS cache
*   Trying 127.0.0.1...
* Connected to 127.0.0.1 (127.0.0.1) port 4243 (#0)
> GET /v1.24/info HTTP/1.1
> User-Agent: curl/7.35.0
> Host: 127.0.0.1:4243
> Accept: */*
> 
^C
```

#### Command list

This behavior occurs with most, but not all, docker commands, summarized in the following list.

:white_check_mark: = responds as expected

:x: = hangs indefinitely

- :white_check_mark: `docker images`
- :x: `docker info` (or `curl 127.0.0.1:4243/v1.24/info`)
- :x: `docker inspect`
- :x: `docker kill`
- :white_check_mark: `docker ps` (or `/v1.24/containers/json`)
- :x: `docker ps --size`
- :x: `docker pull`
- :x: `docker rm`
- :x: `docker rmi`
- :x: `docker run`
- :x: `docker stats`
- :white_check_mark: `docker version`

For debug output of these commands, see **Docker debug logs** below.

**Note:** We recently upgraded from `17.03.1-ce` to `17.09.0-ce`. We had almost the exact same issue before upgrading, with one difference: the **`docker ps` command would hang on 17.03, but does not hang on 17.09.**

##### `inspect`

`inspect` hangs if a container exists:

```
# curl 127.0.0.1:4243/v1.24/containers/136e79f1cbbc/json
^C
```

But if a container doesn't exist, a 404 is returned immediately:
```
root@i-0a7f562-production-2-worker-com-ec2:~# curl 127.0.0.1:4243/v1.24/containers/foo/json
{"message":"No such container: foo"}
```

##### `docker ps --size`

`docker ps` works, but `docker ps -s` hangs indefinitely.

On healthy instances, `docker ps -s` takes a really long time, but eventually succeeds (and given the size of our images, `--size` is expected to be slow):

```
CONTAINER ID        IMAGE                    COMMAND                  CREATED             STATUS                  PORTS               NAMES                                                SIZE
acada0618959        6e0174c4af10             "/sbin/init"             2 minutes ago       Up Less than a second   22/tcp              travis-job-everypoliticia-mexico-diputad-296266035   87.2MB (virtual 9.46GB)
c93285d1fee5        6e0174c4af10             "/sbin/init"             3 minutes ago       Up 41 seconds           22/tcp              travis-job-nexmo-nexmo-develope-296265796            424MB (virtual 9.79GB)
b4bb83d89226        6e0174c4af10             "/sbin/init"             4 minutes ago       Up About a minute       22/tcp              travis-job-mosra-corrade-296265498                   233MB (virtual 9.6GB)
a7cd73d66aa0        6e0174c4af10             "/sbin/init"             4 minutes ago       Up About a minute       22/tcp              travis-job-pecan-pecan-296263555                     402MB (virtual 9.77GB)
d5b177eb1bbc        6e0174c4af10             "/sbin/init"             4 minutes ago       Up 2 minutes            22/tcp              travis-job-chrisburnell-chrisburnell-c-296260645     556MB (virtual 9.93GB)
a23e8b67cb32        6e0174c4af10             "/sbin/init"             6 minutes ago       Up 3 minutes            22/tcp              travis-job-klauspost-pgzip-296247763                 0B
846860047edc        6e0174c4af10             "/sbin/init"             6 minutes ago       Up 3 minutes            22/tcp              travis-job-elgg-elgg-296263457                       366MB (virtual 9.74GB)
324a18dc2bf7        6e0174c4af10             "/sbin/init"             6 minutes ago       Up 3 minutes            22/tcp              travis-job-csound-csound-296264226                   0B
4b69b3aac49b        6e0174c4af10             "/sbin/init"             6 minutes ago       Up 4 minutes            22/tcp              travis-job-samvera-hyrax-296264048                   1e+03MB (virtual 10.4GB)
25d6885ae0e9        6e0174c4af10             "/sbin/init"             12 minutes ago      Up 9 minutes            22/tcp              travis-job-flutter-flutter-296261498                 1.41GB (virtual 10.8GB)
ea3d4d67961d        6e0174c4af10             "/sbin/init"             22 minutes ago      Up 19 minutes           22/tcp              travis-job-angular-angular-296250866                 0B
bbbec401b4e7        6e0174c4af10             "/sbin/init"             32 minutes ago      Up 29 minutes           22/tcp              travis-job-gleninnesy6d6-legana-296249384            235MB (virtual 9.61GB)
f62093c97692        6e0174c4af10             "/sbin/init"             34 minutes ago      Up 32 minutes           22/tcp              travis-job-geraldtonobww-croydonnorth-296244666      235MB (virtual 9.61GB)
3bbc83a6576b        travisci/worker:v3.3.1   "/docker-entrypoin..."   2 hours ago         Up 2 hours                                  travis-worker                                        0B (virtual 30.7MB)
```

## Docker debug logs

Full debug log output from two unhealthy instances:

- [unhealthy instance 1](https://gist.github.com/soulshake/bc1e7162aa75abe692339dee4c210d51)
- [unhealthy instance 2](https://gist.github.com/soulshake/f1d5c957166b7f36b7409013fde68944)

You'll notice a lot of `GET /v1.24/exec/<cid>/json` in the docker debug logs linked above. The worker runs an [`exec inspect`](https://github.com/travis-ci/worker/blob/master/backend/docker.go#L652-L670) every 500ms on each container. (I would not be surprised if this were related somehow.)

Here is the output of various docker commands as they appear in the debug logs in real time:

### Command: `docker -v ps`

Debug output: 
```
time="2017-11-04T22:41:30.800205214Z" level=debug msg="Calling GET /_ping" 
time="2017-11-04T22:41:30.800720080Z" level=debug msg="Calling GET /v1.32/containers/json" 
```

### Command: `docker -v inspect 5308`

Debug output:
```
time="2017-11-04T22:42:25.395579145Z" level=debug msg="Calling GET /_ping" 
time="2017-11-04T22:42:25.396094021Z" level=debug msg="Calling GET /v1.32/containers/5308/json" 
time="2017-11-04T22:42:25.397036385Z" level=debug msg="Calling GET /v1.32/images/5308/json" 
time="2017-11-04T22:42:25.398120304Z" level=debug msg="Calling GET /v1.32/networks/5308" 
time="2017-11-04T22:42:25.398728651Z" level=debug msg="Calling GET /v1.32/volumes/5308" 
time="2017-11-04T22:42:25.398774377Z" level=debug msg="Probing all drivers for volume with name: 5308" 
time="2017-11-04T22:42:25.399702881Z" level=debug msg="Calling GET /v1.32/info" 
```

### Command: `docker -v pull alpine`

Terminal output: 
```
Using default tag: latest
^C
```
Debug output:
```
time="2017-11-04T22:43:22.478701384Z" level=debug msg="Calling GET /_ping" 
time="2017-11-04T22:43:22.479421512Z" level=debug msg="Calling GET /v1.32/info" 
```

## `ps faux` output

- [7 unhealthy instances](https://gist.github.com/soulshake/641802ab9971e8825824f9631ab15ac1)

## `dmesg` output

See [this gist](https://gist.github.com/soulshake/447cf69144f8bd98dcd5eff4ce8345fb) for `dmesg` output from 37 affected instances.

## `fuser -vu /var/run/docker.sock` output

Looks the same on both healthy and unhealthy instances:

```
root@i-0316580-production-2-worker-org-ec2:~# fuser -vu /var/run/docker.sock
                     USER        PID ACCESS COMMAND
/run/docker.sock:    root       2139 F.... (root)dockerd
```

## `lsof /var/run/docker.sock` output

Unhealthy instance:

```
root@i-0901dbc-production-2-worker-org-ec2:~# lsof /var/run/docker.sock
COMMAND  PID USER   FD   TYPE             DEVICE SIZE/OFF   NODE NAME
dockerd 2145 root    5u  unix 0xffff8dcb239ce000      0t0  46283 /var/run/docker.sock
dockerd 2145 root   16u  unix 0xffff8dca5ee3f000      0t0   3252 /var/run/docker.sock
dockerd 2145 root   17u  unix 0xffff8dd2a8d2cc00      0t0  39119 /var/run/docker.sock
dockerd 2145 root   18u  unix 0xffff8dcb26397000      0t0  41132 /var/run/docker.sock
dockerd 2145 root  622u  unix 0xffff8dd2a709d800      0t0 554335 /var/run/docker.sock
dockerd 2145 root  626u  unix 0xffff8dd099d68c00      0t0 571451 /var/run/docker.sock
dockerd 2145 root  653u  unix 0xffff8dca56292400      0t0 431867 /var/run/docker.sock
dockerd 2145 root  685u  unix 0xffff8dcaaedca800      0t0 567658 /var/run/docker.sock
dockerd 2145 root  686u  unix 0xffff8dca2c9da400      0t0 571499 /var/run/docker.sock
dockerd 2145 root  688u  unix 0xffff8dd0ca2b6800      0t0 571501 /var/run/docker.sock
dockerd 2145 root  689u  unix 0xffff8dd2a9d7c000      0t0 571502 /var/run/docker.sock
dockerd 2145 root  690u  unix 0xffff8dcaaedca400      0t0 571503 /var/run/docker.sock
dockerd 2145 root  691u  unix 0xffff8dd2a7099000      0t0 571504 /var/run/docker.sock
dockerd 2145 root  692u  unix 0xffff8dca532d6800      0t0 571505 /var/run/docker.sock
dockerd 2145 root  693u  unix 0xffff8dca2c9df400      0t0 571506 /var/run/docker.sock
dockerd 2145 root  694u  unix 0xffff8dd0f0510c00      0t0 571507 /var/run/docker.sock
dockerd 2145 root  695u  unix 0xffff8dcaaedcdc00      0t0 571508 /var/run/docker.sock
dockerd 2145 root  696u  unix 0xffff8dd0ca2b7c00      0t0 571509 /var/run/docker.sock
dockerd 2145 root  697u  unix 0xffff8dd2a9d7e000      0t0 571510 /var/run/docker.sock
dockerd 2145 root  698u  unix 0xffff8dca1d3c7400      0t0 513936 /var/run/docker.sock
dockerd 2145 root  699u  unix 0xffff8dc991c8a400      0t0 513937 /var/run/docker.sock
```

Healthy instance:

```
root@i-0166d6d-production-2-worker-org-ec2:~# lsof /var/run/docker.sock
COMMAND     PID USER   FD   TYPE             DEVICE SIZE/OFF     NODE NAME
dockerd    2138 root    5u  unix 0xffff971be5d30000      0t0    17307 /var/run/docker.sock
dockerd    2138 root   16u  unix 0xffff971be71ae000      0t0    41184 /var/run/docker.sock
dockerd    2138 root   17u  unix 0xffff971463840000      0t0    40573 /var/run/docker.sock
dockerd    2138 root   18u  unix 0xffff971317917000      0t0     4608 /var/run/docker.sock
docker-co 33764 root   20u  unix 0xffff971a69f06400      0t0 16726937 /tmp/pty062271588/pty.sock
docker-co 96269 root   20u  unix 0xffff9713382e3c00      0t0  8828942 /tmp/pty952192672/pty.sock
```

### `docker info` output

On a healthy instance:

```
Containers: 16
 Running: 16
 Paused: 0
 Stopped: 0
Images: 3
Server Version: 17.09.0-ce
Storage Driver: devicemapper
 Pool Name: docker-202:16-2736138-pool
 Pool Blocksize: 65.54kB
 Base Device Size: 20.4GB
 Backing Filesystem: xfs
 Data file: /dev/direct-lvm/data
 Metadata file: /dev/direct-lvm/metadata
 Data Space Used: 33.67GB
 Data Space Total: 341.4GB
 Data Space Available: 307.7GB
 Metadata Space Used: 37.52MB
 Metadata Space Total: 2.147GB
 Metadata Space Available: 2.11GB
 Thin Pool Minimum Free Space: 34.14GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Library Version: 1.02.77 (2012-10-15)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 06b9cb35161009dcb7123345749fef02f7cea8e0
runc version: 3f2f8b84a77f73d38244dd690525642a72156c64
init version: 949e6fa
Security Options:
 userns
Kernel Version: 4.11.6-041106-generic
Operating System: Ubuntu 14.04.5 LTS
OSType: linux
Architecture: x86_64
CPUs: 32
Total Memory: 58.96GiB
Name: i-04dc98a-production-2-worker-org-ec2.travisci.net
ID: MT5T:YC4V:QEMR:5PXE:XTNI:OLVU:KWAI:RNXG:RASX:WR4Q:ZLDR:RVVU
Docker Root Dir: /mnt/docker/886432.886432
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 162
 Goroutines: 176
 System Time: 2017-11-02T13:47:08.684595977Z
 EventsListeners: 0
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 10.0.0.0/8
 127.0.0.0/8
Live Restore Enabled: false
```

(On an unhealthy instance, `docker info` hangs indefinitely.)

#### lsb_release -a
```
No LSB modules are available.
Distributor ID: Ubuntu
Description:  Ubuntu 14.04.5 LTS
Release:  14.04
Codename: trusty
```

#### uname -a
```
Linux i-0901dbc-production-2-worker-org-ec2.travisci.net 4.11.6-041106-generic #201706170517 SMP Sat Jun 17 09:18:46 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
```

#### docker version

```
Client:
 Version:      17.09.0-ce
 API version:  1.32
 Go version:   go1.8.3
 Git commit:   afdb6d4
 Built:        Tue Sep 26 22:42:38 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.09.0-ce
 API version:  1.32 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   afdb6d4
 Built:        Tue Sep 26 22:41:20 2017
 OS/Arch:      linux/amd64
 Experimental: false
```
#### daemon configuration

```
# cat /etc/docker/daemon-direct-lvm.json 
{
  "graph": "/mnt/docker",
  "hosts": [
    "tcp://127.0.0.1:4243",
    "unix:///var/run/docker.sock"
  ],
  "icc": false,
  "insecure-registries": [
    "10.0.0.0/8"
  ],
  
  "storage-driver": "devicemapper",
  "storage-opts": [
    "dm.basesize=19G",
    "dm.datadev=/dev/direct-lvm/data",
    "dm.metadatadev=/dev/direct-lvm/metadata",
    "dm.fs=xfs"
  ],
  "userns-remap": "default"
}
```

(`"debug": true` has been added for the purposes of this report.)

#### Number of containers

```
# docker ps -aq | wc -l
16
# docker ps --filter status=running | wc -l
16
```

Each instance runs no more than 16 containers at once.

#### Additional environmental details

The instances are running on EC2, instance type c3.8xlarge.

**Comment by thaJeztah** [comment]
ping @tonistiigi @cpuguy83

**Comment by thaJeztah** [comment]
I do see some OOM messages in https://gist.github.com/soulshake/447cf69144f8bd98dcd5eff4ce8345fb

**Comment by cpuguy83** [comment]
Would you be able to provide a stack trace from these stuck instances?
If debug mode is enabled you can `curl <host>/debug/pprof/goroutine?debug=2` or you can send SIGUSR1 to the dockerd process. There will be an entry in the dockerd logs with the location of the file where the stack trace was dumped.

**Comment by sylvainmouquet** [comment]
Do you have a full memory on the host machine ?

**Comment by soulshake** [comment]
@cpuguy83 

[pprof-healthy.log](https://github.com/moby/moby/files/1444110/pprof-healthy.log)
[pprof-unhealthy.log](https://github.com/moby/moby/files/1444111/pprof-unhealthy.log)

**Comment by soulshake** [comment]
@sylvainmouquet 

All containers except one (the "orchestrating" container, which is not limited) are limited to 4G of RAM and sometimes hit OOM. The machines shouldn't be overcommitted because they only run 16 containers at a time (15 "jobs", 1 "orchestrator") and have 58G of RAM total.

Here are some memory stats:

### Unhealthy instance
```
# free -hm
             total       used       free     shared    buffers     cached
Mem:           58G        27G        31G        38M       153M        21G
-/+ buffers/cache:       6.0G        52G
Swap:           0B         0B         0B
```
```
# vmstat -s
     61826972 K total memory
     28741588 K used memory
      8180836 K active memory
     14640680 K inactive memory
     33085384 K free memory
       157248 K buffer memory
     22244132 K swap cache
            0 K total swap
            0 K used swap
            0 K free swap
      3072747 non-nice user cpu ticks
      1473971 nice user cpu ticks
       346513 system cpu ticks
    177708700 idle cpu ticks
       222101 IO-wait cpu ticks
            0 IRQ cpu ticks
         7842 softirq cpu ticks
        51953 stolen cpu ticks
     56044319 pages paged in
    149309004 pages paged out
            0 pages swapped in
            0 pages swapped out
    161978879 interrupts
    227134080 CPU context switches
   1509834036 boot time
       687373 forks

```
```
# cat /proc/meminfo
MemTotal:       61826972 kB
MemFree:        33088700 kB
MemAvailable:   59111000 kB
Buffers:          157280 kB
Cached:         22244140 kB
SwapCached:            0 kB
Active:          8175564 kB
Inactive:       14640668 kB
Active(anon):     425280 kB
Inactive(anon):    28784 kB
Active(file):    7750284 kB
Inactive(file): 14611884 kB
Unevictable:           0 kB
Mlocked:               0 kB
SwapTotal:             0 kB
SwapFree:              0 kB
Dirty:               124 kB
Writeback:             0 kB
AnonPages:        389292 kB
Mapped:           173296 kB
Shmem:             39252 kB
Slab:            4626812 kB
SReclaimable:    4122536 kB
SUnreclaim:       504276 kB
KernelStack:       17440 kB
PageTables:        13848 kB
NFS_Unstable:          0 kB
Bounce:                0 kB
WritebackTmp:          0 kB
CommitLimit:    30913484 kB
Committed_AS:    3852764 kB
VmallocTotal:   34359738367 kB
VmallocUsed:           0 kB
VmallocChunk:          0 kB
HardwareCorrupted:     0 kB
AnonHugePages:    206848 kB
ShmemHugePages:        0 kB
ShmemPmdMapped:        0 kB
CmaTotal:              0 kB
CmaFree:               0 kB
HugePages_Total:       0
HugePages_Free:        0
HugePages_Rsvd:        0
HugePages_Surp:        0
Hugepagesize:       2048 kB
DirectMap4k:      536576 kB
DirectMap2M:    62509056 kB
```
```
# top
top - 14:21:03 up 16:00,  1 user,  load average: 0.00, 0.00, 0.00
Tasks: 601 total,   1 running, 600 sleeping,   0 stopped,   0 zombie
%Cpu(s):  1.7 us,  0.2 sy,  0.8 ni, 97.2 id,  0.1 wa,  0.0 hi,  0.0 si,  0.0 st
KiB Mem:  61826972 total, 28736420 used, 33090552 free,   157396 buffers
KiB Swap:        0 total,        0 used,        0 free. 22244252 cached Mem

   PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND                                                                                                                                                                   
 39892 root      20   0   24104   3396   2576 R   6.2  0.0   0:00.01 top                                                                                                                                                                       
     1 root      20   0   33636   4144   2716 S   0.0  0.0   0:18.99 init                                                                                                                                                                      
     2 root      20   0       0      0      0 S   0.0  0.0   0:00.34 kthreadd                                                                                                                                                                  
     4 root       0 -20       0      0      0 S   0.0  0.0   0:03.34 kworker/0:0H                                                                                                                                                              
     7 root       0 -20       0      0      0 S   0.0  0.0   0:00.00 mm_percpu_wq                                                                                                                                                              
     8 root      20   0       0      0      0 S   0.0  0.0   0:00.25 ksoftirqd/0                                                                                                                                                               
     9 root      20   0       0      0      0 S   0.0  0.0   0:07.12 rcu_sched                                                                                                                                                                 
    10 root      20   0       0      0      0 S   0.0  0.0   0:00.00 rcu_bh                                                                                                                                                                    
    11 root      rt   0       0      0      0 S   0.0  0.0   0:00.16 migration/0                                                                                                                                                               
    12 root      rt   0       0      0      0 S   0.0  0.0   0:00.04 watchdog/0                                                                                                                                                                
    13 root      20   0       0      0      0 S   0.0  0.0   0:00.00 cpuhp/0                                                                                                                                                                   
    14 root      20   0       0      0      0 S   0.0  0.0   0:00.00 cpuhp/1                                                                                                                                                                   
    15 root      rt   0       0      0      0 S   0.0  0.0   0:00.06 watchdog/1                                                                                                                                                                
    16 root      rt   0       0      0      0 S   0.0  0.0   0:00.38 migration/1                                                                                                                                                               
    17 root      20   0       0      0      0 S   0.0  0.0   0:00.33 ksoftirqd/1                                                                                                                                                               
```

Here's a summary on a few unhealthy instances (`free -hm | grep Mem`):

```
             total       used       free     shared    buffers     cached
Mem:           58G        15G        43G        52M       199M       9.9G
Mem:           58G        14G        44G        37M       135M       9.9G
Mem:           58G        27G        31G        72M        69M        21G
Mem:           58G        19G        39G        93M        28M        13G
Mem:           58G        27G        31G       103M        47M        21G
Mem:           58G        20G        38G        72M        29M        15G
Mem:           58G        27G        31G        38M       154M        21G
```

### Healthy instance
```
# free -hm
             total       used       free     shared    buffers     cached
Mem:           58G       4.7G        54G        31M        95M       2.7G
-/+ buffers/cache:       1.9G        57G
Swap:           0B         0B         0B
```

```
# cat /proc/meminfo
MemTotal:       61826972 kB
MemFree:        56668716 kB
MemAvailable:   59981608 kB
Buffers:           97464 kB
Cached:          2890512 kB
SwapCached:            0 kB
Active:          2027640 kB
Inactive:        1632820 kB
Active(anon):     675008 kB
Inactive(anon):    29860 kB
Active(file):    1352632 kB
Inactive(file):  1602960 kB
Unevictable:           0 kB
Mlocked:               0 kB
SwapTotal:             0 kB
SwapFree:              0 kB
Dirty:             34700 kB
Writeback:             0 kB
AnonPages:        637020 kB
Mapped:           240920 kB
Shmem:             32360 kB
Slab:            1055996 kB
SReclaimable:     819672 kB
SUnreclaim:       236324 kB
KernelStack:       18864 kB
PageTables:        13808 kB
NFS_Unstable:          0 kB
Bounce:                0 kB
WritebackTmp:          0 kB
CommitLimit:    30913484 kB
Committed_AS:    4736592 kB
VmallocTotal:   34359738367 kB
VmallocUsed:           0 kB
VmallocChunk:          0 kB
HardwareCorrupted:     0 kB
AnonHugePages:    229376 kB
ShmemHugePages:        0 kB
ShmemPmdMapped:        0 kB
CmaTotal:              0 kB
CmaFree:               0 kB
HugePages_Total:       0
HugePages_Free:        0
HugePages_Rsvd:        0
HugePages_Surp:        0
Hugepagesize:       2048 kB
DirectMap4k:      784384 kB
DirectMap2M:    62261248 kB
```

```
# vmstat -s
     61826972 K total memory
      3679424 K used memory
      1275268 K active memory
      1193292 K inactive memory
     58147548 K free memory
        97520 K buffer memory
      1706156 K swap cache
            0 K total swap
            0 K used swap
            0 K free swap
      7224914 non-nice user cpu ticks
           14 nice user cpu ticks
      1036765 system cpu ticks
     70346629 idle cpu ticks
       210776 IO-wait cpu ticks
            0 IRQ cpu ticks
        25772 softirq cpu ticks
        59141 stolen cpu ticks
    151561857 pages paged in
    536935540 pages paged out
            0 pages swapped in
            0 pages swapped out
    610682678 interrupts
    928602522 CPU context switches
   1509866707 boot time
      2317501 forks
```

```
# top
top - 14:20:33 up  6:55,  1 user,  load average: 7.06, 17.25, 12.80
Tasks: 524 total,   3 running, 521 sleeping,   0 stopped,   0 zombie
%Cpu(s):  8.9 us,  0.6 sy,  0.0 ni, 89.9 id,  0.6 wa,  0.0 hi,  0.0 si,  0.0 st
KiB Mem:  61826972 total,  5153648 used, 56673324 free,    97580 buffers
KiB Swap:        0 total,        0 used,        0 free.  2230752 cached Mem

   PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND                                                                                                                                                                   
104375 888432    20   0 3916828 950088  24092 S 198.5  1.5   0:42.70 java                                                                                                                                                                      
 97640 888432    20   0  466080 205900  21228 R  87.6  0.3   1:01.75 php                                                                                                                                                                       
 96298 888432    20   0  297020  90476  18252 S  12.6  0.1   0:01.46 mongod                                                                                                                                                                    
125044 root      20   0       0      0      0 S   2.7  0.0   0:14.57 kworker/u256:0                                                                                                                                                            
 86442 root      20   0       0      0      0 S   2.3  0.0   0:27.92 kworker/u257:1                                                                                                                                                            
119794 root      20   0       0      0      0 S   1.3  0.0   0:00.30 kworker/8:3                                                                                                                                                               
  2136 root      20   0 6252292 110948  29984 S   0.7  0.2  23:25.41 dockerd                                                                                                                                                                   
  3984 root      20   0   45744  32684  10728 S   0.7  0.1 339:10.12 travis-worker                                                                                                                                                             
103024 886542    20   0  272232  54488  21940 S   0.7  0.1   0:00.43 mongod                                                                                                                                                                    
    61 root       0 -20       0      0      0 S   0.3  0.0   0:14.55 kworker/8:0H                                                                                                                                                              
    74 root       0 -20       0      0      0 S   0.3  0.0   0:05.27 kworker/10:0H                                                                                                                                                             
  2508 root       0 -20       0      0      0 S   0.3  0.0   0:12.10 kworker/9:1H                                                                                                                                                              
 68500 root      20   0       0      0      0 S   0.3  0.0   0:01.56 kworker/u256:3                                                                                                                                                            
 76226 886432    20   0   19716   2528   1848 S   0.3  0.0   0:00.11 irqbalance                                                                                                                                                                
104401 root      20   0   24092   3460   2532 R   0.3  0.0   0:00.08 top                                                                                                                                                                       
```

Here's a summary on healthy instances: (`free -hm | grep Mem`):

```
             total       used       free     shared    buffers     cached
Mem:           58G       5.8G        53G        39M        98M       3.0G
Mem:           58G       5.8G        53G        39M        98M       3.0G
Mem:           58G       9.3G        49G        99M       198M       4.9G
Mem:           58G       9.4G        49G        99M       198M       4.9G
Mem:           58G        17G        41G       213M       124M       8.6G
Mem:           58G        17G        41G       213M       124M       8.7G
Mem:           58G        39G        19G       292M       307M        22G
Mem:           58G        39G        19G       292M       307M        22G
Mem:           58G        26G        32G        72M        32M        20G
Mem:           58G        26G        32G        72M        32M        20G
Mem:           58G        24G        34G       185M       112M        11G
```

```
# docker stats
CONTAINER           CPU %               MEM USAGE / LIMIT     MEM %               NET I/O             BLOCK I/O           PIDS
b38fd20837e3        129.30%             92.37MiB / 4GiB       2.26%               34.3MB / 131kB      125MB / 254kB       61
08de53538380        92.69%              318.9MiB / 4GiB       7.79%               34.6MB / 206kB      197MB / 627kB       62
ccbeec414536        97.93%              335.5MiB / 4GiB       8.19%               54.9MB / 332kB      202MB / 8.34MB      51
ad1f79806cad        103.79%             360.4MiB / 4GiB       8.80%               45.5MB / 214kB      213MB / 795kB       48
ab9db5bb5aa4        104.68%             573.5MiB / 4GiB       14.00%              88.4MB / 709kB      216MB / 6.18MB      63
c29a3ed77a6a        151.25%             769.1MiB / 4GiB       18.78%              269MB / 1.38MB      243MB / 10.3MB      85
6dbf5ac9bffc        104.75%             878.1MiB / 4GiB       21.44%              158MB / 956kB       266MB / 238MB       80
81e5889ea596        2.44%               34.11MiB / 58.96GiB   0.06%               21.2MB / 120MB      19.2MB / 0B         71
^C
```

(`81e5889ea596` is the "orchestrator" container I mentioned above.)

**Comment by cpuguy83** [comment]
Looks like it's blocked waiting on devicemapper:

<details>

```
goroutine 76199 [syscall, 887 minutes, locked to thread]:
github.com/docker/docker/pkg/devicemapper._Cfunc_dm_udev_wait(0xd4d2022, 0x0)
	github.com/docker/docker/pkg/devicemapper/_obj/_cgo_gotypes.go:387 +0x4d
github.com/docker/docker/pkg/devicemapper.dmUdevWaitFct(0xd4d2022, 0xc4220e2928)
	/go/src/github.com/docker/docker/pkg/devicemapper/devmapper_wrapper.go:227 +0x2a
github.com/docker/docker/pkg/devicemapper.UdevWait(0xc421e67d08, 0x0, 0x0)
	/go/src/github.com/docker/docker/pkg/devicemapper/devmapper.go:257 +0x39
github.com/docker/docker/pkg/devicemapper.activateDevice(0xc422ebd8f0, 0x27, 0xc421c0a7e0, 0x5c, 0x10d, 0x4c0000000, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/docker/docker/pkg/devicemapper/devmapper.go:764 +0x559
github.com/docker/docker/pkg/devicemapper.ActivateDevice(0xc422ebd8f0, 0x27, 0xc421c0a7e0, 0x5c, 0x10d, 0x4c0000000, 0x27, 0xc422e1fa00)
	/go/src/github.com/docker/docker/pkg/devicemapper/devmapper.go:725 +0x79
github.com/docker/docker/daemon/graphdriver/devmapper.(*DeviceSet).activateDeviceIfNeeded(0xc4201983c0, 0xc421718680, 0xc420198300, 0xc421718680, 0x0)
	/go/src/github.com/docker/docker/daemon/graphdriver/devmapper/deviceset.go:533 +0x277
github.com/docker/docker/daemon/graphdriver/devmapper.(*DeviceSet).MountDevice(0xc4201983c0, 0xc423e2ae10, 0x45, 0xc42132fd50, 0x70, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/docker/docker/daemon/graphdriver/devmapper/deviceset.go:2363 +0x156
github.com/docker/docker/daemon/graphdriver/devmapper.(*Driver).Get(0xc42014b020, 0xc423e2ae10, 0x45, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/docker/docker/daemon/graphdriver/devmapper/driver.go:198 +0x40d
github.com/docker/docker/daemon/graphdriver.(*NaiveDiffDriver).Get(0xc420215c00, 0xc423e2ae10, 0x45, 0x0, 0x0, 0xc420b3eaa0, 0x0, 0x0, 0x40c205)
	<autogenerated>:30 +0x75
github.com/docker/docker/layer.(*layerStore).initMount(0xc42014b080, 0xc4217184c0, 0x40, 0xc4228ff5c0, 0x40, 0x0, 0x0, 0xc421e66650, 0x0, 0x1912ce0, ...)
	/go/src/github.com/docker/docker/layer/layer_store.go:655 +0x1e0
github.com/docker/docker/layer.(*layerStore).CreateRWLayer(0xc42014b080, 0xc421533e00, 0x40, 0xc423e2adc0, 0x47, 0xc420b3ea60, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/docker/docker/layer/layer_store.go:527 +0x629
github.com/docker/docker/daemon.(*Daemon).setRWLayer(0xc4204ab400, 0xc422fc2480, 0xc421d05800, 0x0)
	/go/src/github.com/docker/docker/daemon/create.go:269 +0x173
github.com/docker/docker/daemon.(*Daemon).create(0xc4204ab400, 0xc420f86da3, 0x29, 0xc421de6000, 0xc421d05800, 0x0, 0x0, 0x1ab1e03, 0x5, 0x0, ...)
	/go/src/github.com/docker/docker/daemon/create.go:157 +0x43f
github.com/docker/docker/daemon.(*Daemon).containerCreate(0xc4204ab400, 0xc420f86da3, 0x29, 0xc421de6000, 0xc421d05800, 0x0, 0x0, 0x1ab1e03, 0x5, 0x0, ...)
	/go/src/github.com/docker/docker/daemon/create.go:70 +0x1fa
github.com/docker/docker/daemon.(*Daemon).ContainerCreate(0xc4204ab400, 0xc420f86da3, 0x29, 0xc421de6000, 0xc421d05800, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/docker/docker/daemon/create.go:33 +0xa1
github.com/docker/docker/api/server/router/container.(*containerRouter).postContainersCreate(0xc420340740, 0x7f41c3860608, 0xc421680240, 0x2738020, 0xc4219bb500, 0xc420504200, 0xc4225edd40, 0x1ab1b6a, 0x5)
	/go/src/github.com/docker/docker/api/server/router/container/container_routes.go:418 +0x272
github.com/docker/docker/api/server/router/container.(*containerRouter).(github.com/docker/docker/api/server/router/container.postContainersCreate)-fm(0x7f41c3860608, 0xc421680240, 0x2738020, 0xc4219bb500, 0xc420504200, 0xc4225edd40, 0x7f41c3860608, 0xc421680240)
	/go/src/github.com/docker/docker/api/server/router/container/container.go:61 +0x69
github.com/docker/docker/api/server/middleware.ExperimentalMiddleware.WrapHandler.func1(0x7f41c3860608, 0xc421680240, 0x2738020, 0xc4219bb500, 0xc420504200, 0xc4225edd40, 0x7f41c3860608, 0xc421680240)
	/go/src/github.com/docker/docker/api/server/middleware/experimental.go:27 +0xd8
github.com/docker/docker/api/server/middleware.VersionMiddleware.WrapHandler.func1(0x7f41c3860608, 0xc4225eddd0, 0x2738020, 0xc4219bb500, 0xc420504200, 0xc4225edd40, 0x666, 0xc421f6b500)
	/go/src/github.com/docker/docker/api/server/middleware/version.go:58 +0x487
github.com/docker/docker/pkg/authorization.(*Middleware).WrapHandler.func1(0x7f41c3860608, 0xc4225eddd0, 0x2738020, 0xc4219bb500, 0xc420504200, 0xc4225edd40, 0x0, 0x756ea166b97d38)
	/go/src/github.com/docker/docker/pkg/authorization/middleware.go:59 +0x871
github.com/docker/docker/api/server/middleware.DebugRequestMiddleware.func1(0x7f41c3860608, 0xc4225eddd0, 0x2738020, 0xc4219bb500, 0xc420504200, 0xc4225edd40, 0x7f41c3860608, 0xc4225eddd0)
	/go/src/github.com/docker/docker/api/server/middleware/debug.go:53 +0x527
github.com/docker/docker/api/server.(*Server).makeHTTPHandler.func1(0x2738020, 0xc4219bb500, 0xc420504200)
	/go/src/github.com/docker/docker/api/server/server.go:137 +0x21a
net/http.HandlerFunc.ServeHTTP(0xc420b04ce0, 0x2738020, 0xc4219bb500, 0xc420504200)
	/usr/local/go/src/net/http/server.go:1942 +0x44
github.com/docker/docker/vendor/github.com/gorilla/mux.(*Router).ServeHTTP(0xc4202bc280, 0x2738020, 0xc4219bb500, 0xc420504200)
	/go/src/github.com/docker/docker/vendor/github.com/gorilla/mux/mux.go:103 +0x255
github.com/docker/docker/api/server.(*routerSwapper).ServeHTTP(0xc420c37870, 0x2738020, 0xc4219bb500, 0xc420504200)
	/go/src/github.com/docker/docker/api/server/router_swapper.go:29 +0x70
net/http.serverHandler.ServeHTTP(0xc4200b4bb0, 0x2738020, 0xc4219bb500, 0xc420504200)
	/usr/local/go/src/net/http/server.go:2568 +0x92
net/http.(*conn).serve(0xc420ae1180, 0x2739c20, 0xc4228fee40)
	/usr/local/go/src/net/http/server.go:1825 +0x612
created by net/http.(*Server).Serve
	/usr/local/go/src/net/http/server.go:2668 +0x2ce
```

</details>

Given that the goroutine has been waiting for over 14 hours, I don't think a response is coming.
This could be an issue with the 4.11 kernel. I'd recommend only using an LTS kernel.
I noticed the dm library version is quite old as well.

I have a pretty strong feeling that this combination of the non-stock (and brand new) kernel with Ubuntu 14.04 packages is not a good mix.

**Comment by soulshake** [comment]
Having just come across [this issue](https://github.com/moby/moby/issues/33603), I  now find that running `ipcs` reveals something about the problem.

- on unhealthy hosts, the output looks like this:

```
------ Shared Memory Segments --------
key        shmid      owner      perms      bytes      nattch     status      

------ Semaphore Arrays --------
key        semid      owner      perms      nsems     
0x0d4dbc91 378044416  root       600        1         <========

------ Message Queues --------
key        msqid      owner      perms      used-bytes   messages    
```

- on healthy hosts, like this:
```
------ Shared Memory Segments --------
key        shmid      owner      perms      bytes      nattch     status      

------ Semaphore Arrays --------
key        semid      owner      perms      nsems     

------ Message Queues --------
key        msqid      owner      perms      used-bytes   messages    
```

**Running `echo 'y' | dmsetup udevcomplete_all` immediately brings the engine back to life** (though I'm not sure for how long, because in our case we've already marked the instance for termination).

FYI:

```
# cat /proc/sys/kernel/sem
32000	1024000000	500	32000
```

By the way, we did switch to the most recent LTS kernel (`4.9.6`) as @cpuguy83 suggested; the issue still occurs.

**Comment by soulshake** [comment]
`dockerd` seems to be the process responsible for creating this semaphor:

```
# ipcs -s
------ Semaphore Arrays --------
key        semid      owner      perms      nsems     
0x0d4d38ba 225411072  root       600        1         
```

More info on `semid` `225411072`:

```
# ipcs -s -i 225411072
Semaphore Array semid=225411072
uid=0     gid=0     cuid=0     cgid=0
mode=0600, access_perms=0600
nsems = 1
otime = Thu Nov  9 23:45:29 2017  
ctime = Thu Nov  9 23:45:29 2017  
semnum     value      ncount     zcount     pid       
0          1          0          1          2140      
```
Then on `pid` `2140`:
```# ps -p 2140
   PID TTY          TIME CMD
  2140 ?        00:47:11 dockerd
```

Since there doesn't seem to be any reference to explicit cookie handling in the moby code (please correct me if I'm wrong), is it reasonable to assume that this is set by the `libdevmapper` code that the `devmapper` storage driver is using?

**Comment by sean-abbott** [comment]
I am having exactly this issue with this docker version, on centos with devicemapper in AWS.

Running the commands suggested by @soulshake cleared the issue for me. We're planning on setting higher memory allowances and running more instances to try and avoid running into this, but a solution that allowed us to run out of memory without freezing docker via sticking semaphore would be nice.

**Comment by sean-abbott** [comment]
Is this issue fixed in versions past 17.09.0-ce?

**Comment by thaJeztah** [comment]
Let me close this ticket for now, as it looks like it went stale.
